### ISSUE TITLE
Windows support

### ISSUE THREAD

**Issue by johannbrehmer** [MilesCranmer/PySR#17]
Hi Miles,

first of all, this is awesome. Thanks so much for making this.

A student I'm working with is trying to run PySR under Windows. Is that in principle supported?

PySR's dependencies don't seem to have any issues with Windows, but `pysr.pysr` throws a `FileNotFoundError` when accessing `/tmp/.hyperparams_{rand_string}.hl'`. Seems to be because of the different file system structure under Windows. If this is the only issue, how would you feel about using something like [tempfile](https://docs.python.org/3/library/tempfile.html) to generate temporary files in a more OS-independent way?

I am happy to try this and open a PR once it works.

Cheers,
Johann

**Comment by johannbrehmer** [comment]
I'm testing platform-independent paths at https://github.com/johannbrehmer/PySR/tree/tempfile . That seems to work smoothly under Mac OS, I'm waiting to hear back if it works under Windows as well.

**Comment by MilesCranmer** [comment]
Thanks @johannbrehmer! This is definitely an important issue.

I helped someone over email solve this; but I didn't hear back if the final change was successful or not. Here is the things we had to change:

- Changed `/tmp/` to the windows equivalent, I think `%TEMP%`?
   - However, your solution is much better! The `tempfile` package looks great - thanks for bringing this to my attention. I had a quick look through your code and it looks like a great way of implementing it. Feel free to submit a PR.
- The other thing that needs to be changed is `subprocess.Popen` expects a list on POSIX-compliant systems, but a single string instead on windows. So an if-statement on windows would call `command = ' '.join(command)`. 

Other than this, I think we would be good to go on Windows. Let me know if this works for your student!

Thanks again for contributing to PySR!
Cheers,
Miles

**Comment by johannbrehmer** [comment]
Hi Miles,

Thanks for your answer! I had no idea that `subprocess.Popen` behaves differently depending on the system. I'll use your proposed solution for that and open a PR.

Cheers,
Johann

**Comment by MilesCranmer** [comment]
Awesome, sounds good!

Actually I just re-read the docs part about this: https://docs.python.org/3/library/subprocess.html#converting-argument-sequence. It might actually auto-convert the list to a string in the most recent version of Python? But we can see if it works!

**Comment by MilesCranmer** [comment]
Also, it might be useful to have a PySR option to turn off the automatic `tempfile` context manager cleanup (for debugging, or running code directly from Julia). It looks like `delete=False` might work for some of the objects?

**Comment by johannbrehmer** [comment]
Running the code with the unmodified `subprocess.Popen(command, ...)` on Windows results in a FileNotFountError. I haven't tested your `' '.join(command)` solution yet, but I convinced my partner to let me use her Windows laptop for testing later, so I'll get to that soon.

**Comment by johannbrehmer** [comment]
Good point about having the option to not delete the temporary folder. Sadly, that doesn't seem to be an option with `tempfile.TemporaryDirectory`. Maybe then we should switch to a set of a few `tempfile.NamedTemporaryFile` objects (which do have a `delete` keyword like you said).

**Comment by johannbrehmer** [comment]
Option A, where the temp files might not be in the same directory: 
```
    with \
        tempfile.NamedTemporaryFile(suffix=f'.hyperparams_{rand_string}.jl', dir=tempdir, delete=delete_tempfiles) as hyperparam_file, \
        tempfile.NamedTemporaryFile(suffix=f'.dataset_{rand_string}.jl', dir=tempdir, delete=delete_tempfiles) as dataset_file, \
        tempfile.NamedTemporaryFile(suffix=f'.runfile_{rand_string}.jl', dir=tempdir, delete=delete_tempfiles) as run_file \
    :
    # ...
```

Option B, which seems less pythonic:
```
tmpdir = Path(tempfile.mkdtemp(dir=tempdir))
# ...
if delete_tempfiles:
    shutil.rmtree(tmpdir)
```

**Comment by johannbrehmer** [comment]
Option A might not be good for Windows because

> Whether the name can be used to open the file a second time, while the named temporary file is still open, varies across platforms (it can be so used on Unix; it cannot on Windows NT or later)

(from the [`tempfile.NamedTemporaryFile` docs](https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile))

**Comment by MilesCranmer** [comment]
Both look okay to me! Whichever you decide on

**Comment by johannbrehmer** [comment]
Fun little issue that took me embarrassingly long to figure out: Windows filenames contain backslashes. PySR writes filenames into a Julia script. Backslashes are escape characters in Julia strings. Boom

**Comment by MilesCranmer** [comment]
Awesome! 

Btw, I just googled this and apparently another way to do this in Julia is with the "raw" command:
```julia
raw"C:\Users\MyUser\Folder\File.jl"
```

**Comment by johannbrehmer** [comment]
Ah, interesting, that might be even better as long as there are no other `\n` or similar in these strings (I don't think there are).

**Comment by MilesCranmer** [comment]
Thanks again! Merged #18.

**Comment by MilesCranmer** [comment]
Update: travis-ci is now doing continuous testing for Windows/macOS as well!
